### PR TITLE
Fix build break by bumping DependencyModel package version

### PR DESCRIPTION
**Pull request by eerhardt**
Fixes build errors like the following:

```
src\libraries\pkg\test\testPackages.proj(217,5): error : (NETCORE_ENGINEERING_TELEMETRY=Build) Validation Failed: Microsoft.Extensions.DependencyModel is harvesting assets from package version 3.1.3 which is not the latest for that package release. Latest package version from that release is 3.1.4. In order to fix this, run `dotnet msbuild Microsoft.Extensions.DependencyModel.pkgproj /t:UpdateHarvestVersionOnPackageIndex /p:UpdateStablePackageInfo=true`
```

See https://dev.azure.com/dnceng/public/_build/results?buildId=641411&view=logs&j=423ae147-26a3-5a4d-a4d3-4b3fc1dd598e&t=98a4dd5a-baea-5dc4-484b-ccbd315f07c8 for an example.

### CODE REVIEW

File: src/libraries/pkg/baseline/packageIndex.json
Comment (joperezr) [inline src/libraries/pkg/baseline/packageIndex.json:552]: I’m not sure about this one but I understand you didn’t do this manually so I’m ok with it

**Review by joperezr — APPROVED** [review]

File: src/libraries/pkg/baseline/packageIndex.json
Comment (eerhardt) [reply src/libraries/pkg/baseline/packageIndex.json:552]: I **did** do this manually. 😄 

The assembly version in DependencyModel got updated to `3.1.4.0`, so I updated this as well.

Comment (eerhardt) [reply src/libraries/pkg/baseline/packageIndex.json:552]: How do I do it automatically?

**Review by safern — APPROVED** [review]

File: src/libraries/pkg/baseline/packageIndex.json
Comment (ericstj) [reply src/libraries/pkg/baseline/packageIndex.json:552]: You should not update this.  This is meant to indicate the first package version an assembly version appeared in.

Comment (joperezr) [reply src/libraries/pkg/baseline/packageIndex.json:552]: I believe that the break in the build error should tell you the command you have to run in order to fix this automatically

Comment (joperezr) [reply src/libraries/pkg/baseline/packageIndex.json:552]: From the error you copied on the PR description I can see that it is telling you the instructions on how to fix it

Comment (eerhardt) [reply src/libraries/pkg/baseline/packageIndex.json:552]: > You should not update this. This is meant to indicate the first package version an assembly version appeared in.

I'm wondering why this line is even necessary. Can it be removed?

Comment (eerhardt) [reply src/libraries/pkg/baseline/packageIndex.json:552]: Note, if I don't update this line, I get:

```
C:\Users\eerhardt\.nuget\packages\microsoft.dotnet.build.tasks.packaging\5.0.0-beta.20256.5\build\Packaging.targets(1142,5): error : PackageIndex from F:\git\runtime\src\libraries\pkg\baseline\packageIndex.json is missing an assembly version entry for 3.1.4.0 for package Microsoft.Extensions.DependencyModel.  Please run /t:UpdatePackageIndex on this project to commit an update. [F:\git\runtime\src\libraries\Microsoft.Extensions.DependencyModel\pkg\Microsoft.Extensions.DependencyModel.pkgproj]
C:\Users\eerhardt\.nuget\packages\microsoft.dotnet.build.tasks.packaging\5.0.0-beta.20256.5\build\Packaging.targets(1142,5): error : PackageIndex from F:\git\runtime\src\libraries\pkg\baseline\packageIndex.json is has an assembly version entry(s) for 3.1.3.0 which are no longer in package Microsoft.Extensions.DependencyModel.  Please run /t:UpdatePackageIndex on this project to commit an update. [F:\git\runtime\src\libraries\Microsoft.Extensions.DependencyModel\pkg\Microsoft.Extensions.DependencyModel.pkgproj]
```

Comment (joperezr) [reply src/libraries/pkg/baseline/packageIndex.json:552]: We actually do use it to determine which package dependency version other packages will get if they depend on DependencyModel. For example: say you have a library B which depends on DependencyModel, and more specifically they depend on assembly version 3.1.4.0. Our package infrastructure will then come to this table, and check which package version is the one where that assembly version first appeared, (the table above actually looks incorrect as I would assume that it should have something like `"3.1.4.0": "3.1.4"`) and then determine that package B should depend on at least DependencyModel package version 3.1.4. Assuming there is no BaselineVersion for DependencyModel (which in this case we do have one, but let's pretend we don't) then B would depend on 3.1.4 DependencyModel. In this case we do have a BaselineVersion, so even if our package infrastructure chooses package version 3.1.4 for the dependency, it will then baseline up to 5.0.0 because BaselineVersion > InitialChoosenPackageVersion.

Does that sort of make sense?

Comment (joperezr) [reply src/libraries/pkg/baseline/packageIndex.json:552]: FWIW, I believe that the right contents of that dictionary for DependencyModel should look something like:

```json
      "AssemblyVersionInPackageVersion": {
        "3.1.3.0": "3.1.3",
        "3.1.4.0": "3.1.4",
        "5.0.0.0": "5.0.0"
      }
```

Comment (eerhardt) [reply src/libraries/pkg/baseline/packageIndex.json:552]: Since we are so far in the CI now, I want to wait for it to finish and merge this to unblock the build. I can put up another change for the above.

Comment (joperezr) [reply src/libraries/pkg/baseline/packageIndex.json:552]: I'm fine with that as long as we take care of it 😄

**Comment by safern** [comment]
All configurations is green so I'll go ahead and merge to unblock other PR/CI builds.

File: src/libraries/pkg/baseline/packageIndex.json
Comment (eerhardt) [reply src/libraries/pkg/baseline/packageIndex.json:552]: https://github.com/dotnet/runtime/pull/36301

**Comment by trylek** [comment]
I'm seeing a very similar looking issue in my latest PR run:

https://dev.azure.com/dnceng/public/_build/results?buildId=657927&view=logs&jobId=5373b9aa-3658-546d-0a78-c247de617f40&j=5373b9aa-3658-546d-0a78-c247de617f40&t=d7b36ffb-0896-506c-9920-38786e77292f

<pre>
F:\workspace\_work\1\s\src\libraries\pkg\test\testPackages.proj(217,5): error : Validation Failed: Microsoft.Extensions.DependencyModel is harvesting assets from package version 3.1.3 which is not the latest for that package release. Latest package version from that release is 3.1.4. In order to fix this, run `dotnet msbuild Microsoft.Extensions.DependencyModel.pkgproj /t:UpdateHarvestVersionOnPackageIndex /p:UpdateStablePackageInfo=true`
##[error]src\libraries\pkg\test\testPackages.proj(217,5): error : (NETCORE_ENGINEERING_TELEMETRY=Build) Validation Failed: Microsoft.Extensions.DependencyModel is harvesting assets from package version 3.1.3 which is not the latest for that package release. Latest package version from that release is 3.1.4. In order to fix this, run `dotnet msbuild Microsoft.Extensions.DependencyModel.pkgproj /t:UpdateHarvestVersionOnPackageIndex /p:UpdateStablePackageInfo=true`

Build FAILED.
</pre>

Has it somehow regressed again or am I hitting an unrelated issue that just has similar manifestations?

Thanks

Tomas

**Comment by joperezr** [comment]
This symptom can only really be caused by one thing which is your local packageIndex.json is not up to date, but that shouldn’t happen unless you reverted the changes @eerhardt made or you are building an old branch that doesn’t have his change.

**Comment by joperezr** [comment]
From looking at the link you attached, looks like your PR was against a different branch, dev/infrastructure, so it is very possible that his change isn’t present on that branch. In order to fix this, just port his change over there which should fox the issue

**Comment by trylek** [comment]
Ahh, of course you're right, thanks @joperezr for reminding me, it's been a while since I last worked in a feature branch and I completely forgot that. I guess I should follow up with @jashook though as I was under the impression that the <code>dev/infrastructure</code> branch should be on a weekly sync-up plan with master and this fix seems almost two weeks old.

**Comment by ManickaP** [comment]
Seeing something similar in #34948, as well as on one of the newest PR #37662 in repo as of now :
```
src\libraries\pkg\test\testPackages.proj(217,5): error : (NETCORE_ENGINEERING_TELEMETRY=Build) Validation Failed: Microsoft.Extensions.DependencyModel is harvesting assets from package version 3.1.4 which is not the latest for that package release. Latest package version from that release is 3.1.5. In order to fix this, run `dotnet msbuild Microsoft.Extensions.DependencyModel.pkgproj /t:UpdateHarvestVersionOnPackageIndex /p:UpdateStablePackageInfo=true`
```

**Comment by eerhardt** [comment]
@ManickaP - the same thing happened again with `3.1.5`. https://github.com/dotnet/runtime/pull/37663 is fixing it. https://github.com/dotnet/runtime/issues/37661 is tracking fixing the underlying issue

**Pull request merged**
